operator/apis/loki/v1beta1/lokistack_types.go

```diff
@@ -835,3 +835,490 @@ type LokiStackList struct {
+			dst.Spec.Template.Gateway = &v1.LokiComponentSpec{
+				Replicas:     src.Spec.Template.Gateway.Replicas,
+				NodeSelector: src.Spec.Template.Gateway.NodeSelector,
+				Tolerations:  src.Spec.Template.Gateway.Tolerations,
+			}
+		}
+		if src.Spec.Template.IndexGateway != nil {
+			dst.Spec.Template.IndexGateway = &v1.LokiComponentSpec{
+				Replicas:     src.Spec.Template.IndexGateway.Replicas,
+				NodeSelector: src.Spec.Template.IndexGateway.NodeSelector,
+				Tolerations:  src.Spec.Template.IndexGateway.Tolerations,
+			}
+		}
+		if src.Spec.Template.Ruler != nil {
+			dst.Spec.Template.Ruler = &v1.LokiComponentSpec{
+				Replicas:     src.Spec.Template.Ruler.Replicas,
+				NodeSelector: src.Spec.Template.Ruler.NodeSelector,
+				Tolerations:  src.Spec.Template.Ruler.Tolerations,
+			}
+		}
+	}
+
+	if src.Spec.Tenants != nil {
+		dst.Spec.Tenants = &v1.TenantsSpec{
+			Mode: v1.ModeType(src.Spec.Tenants.Mode),
+		}
+
+		for _, srcAuth := range src.Spec.Tenants.Authentication {
+			dstAuth := v1.AuthenticationSpec{
+				TenantName: srcAuth.TenantName,
+				TenantID:   srcAuth.TenantID,
+			}
+
+			if srcAuth.OIDC != nil {
+				dstAuth.OIDC = &v1.OIDCSpec{
+					Secret: &v1.TenantSecretSpec{
+						Name: srcAuth.OIDC.Secret.Name,
+					},
+					IssuerURL:     srcAuth.OIDC.IssuerURL,
+					RedirectURL:   srcAuth.OIDC.RedirectURL,
+					GroupClaim:    srcAuth.OIDC.GroupClaim,
+					UsernameClaim: srcAuth.OIDC.UsernameClaim,
+				}
+			}
+
+			dst.Spec.Tenants.Authentication = append(dst.Spec.Tenants.Authentication, dstAuth)
+		}
+
+		if src.Spec.Tenants.Authorization != nil {
+			dstAuthz := &v1.AuthorizationSpec{}
+
+			if src.Spec.Tenants.Authorization.OPA != nil {
+				dstAuthz.OPA = &v1.OPASpec{
+					URL: src.Spec.Tenants.Authorization.OPA.URL,
+				}
+			}
+
+			for _, srcRole := range src.Spec.Tenants.Authorization.Roles {
+				dstRole := v1.RoleSpec{
+					Name:        srcRole.Name,
+					Resources:   srcRole.Resources,
+					Tenants:     srcRole.Tenants,
+					Permissions: []v1.PermissionType{},
+				}
+
+				for _, perm := range srcRole.Permissions {
+					dstRole.Permissions = append(dstRole.Permissions, v1.PermissionType(perm))
+				}
+
+				dstAuthz.Roles = append(dstAuthz.Roles, dstRole)
+			}
+
+			for _, srcBinding := range src.Spec.Tenants.Authorization.RoleBindings {
+				dstBinding := v1.RoleBindingsSpec{
+					Name:  srcBinding.Name,
+					Roles: srcBinding.Roles,
+				}
+
+				for _, srcSubject := range srcBinding.Subjects {
+					dstBinding.Subjects = append(dstBinding.Subjects, v1.Subject{
+						Name: srcSubject.Name,
+						Kind: v1.SubjectKind(srcSubject.Kind),
+					})
+				}
+
+				dstAuthz.RoleBindings = append(dstAuthz.RoleBindings, dstBinding)
+			}
+
+			dst.Spec.Tenants.Authorization = dstAuthz
+		}
+	}
+
+	return nil
+}
+
+// ConvertFrom converts from the Hub version (v1) to this version (v1beta1).
+// nolint:golint
+func (dst *LokiStack) ConvertFrom(srcRaw conversion.Hub) error {
```

Comment:
Do we even need the `ConvertFrom`? The way I see it, `ConvertTo` is invoked once to update the stored resource, but there is no way of getting a `v1beta1` version out of the apiserver anymore. Don't know if it is possible to just leave it out though.

Reply:
`ConverTo` and `ConvertFrom` are required to suffice the conversion interface.